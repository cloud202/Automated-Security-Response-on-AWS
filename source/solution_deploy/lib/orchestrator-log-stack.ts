#!/usr/bin/env node
/*****************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.        *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import * as cdk from '@aws-cdk/core';
import { LogGroup, CfnLogGroup, RetentionDays } from '@aws-cdk/aws-logs';
import { Key } from '@aws-cdk/aws-kms';
 
export interface OrchLogStackProps extends cdk.StackProps {
    description: string;
    logGroupName: string;
    solutionId: string;
}
  
export class OrchLogStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props: OrchLogStackProps) {
  
    super(scope, id, props);
    const stack = cdk.Stack.of(this);

    const reuseOrchLogGroup = new cdk.CfnParameter(this, 'Reuse Log Group', {
        type: "String",
        description: `Reuse existing Orchestrator Log Group? Choose "yes" if the log group already exists, else "no"`,
        default: "no",
        allowedValues: ["yes", "no"]
    })
    reuseOrchLogGroup.overrideLogicalId(`ReuseOrchestratorLogGroup`)

    const kmsKeyArn = new cdk.CfnParameter(this, 'KMS Key Arn', {
        type: "String",
        description: `ARN of the KMS key to use to encrypt log data.`,
    })
    kmsKeyArn.overrideLogicalId(`KmsKeyArn`)
     
    /**********************
     * Encrypted log group
     */
    // As of March 2021, CWLogs encryption is not yet supported in GovCloud
    // Choose based on partition

    const kmsKey = Key.fromKeyArn(this, "KmsKey", kmsKeyArn.valueAsString)
    const orchestratorLogGroupEncrypted: LogGroup = new LogGroup(this, 'Orchestrator-Logs-Encrypted', {
        logGroupName: props.logGroupName,
        removalPolicy: cdk.RemovalPolicy.RETAIN,
        retention: RetentionDays.ONE_YEAR,
        encryptionKey: kmsKey
    });

    /************************
     * Unencrypted log group
     */
    const orchestratorLogGroupNOTEncrypted: LogGroup = new LogGroup(this, 'Orchestrator-Logs', {
        logGroupName: props.logGroupName,
        removalPolicy: cdk.RemovalPolicy.RETAIN,
        retention: RetentionDays.ONE_YEAR
    });
 
    /*******************
     *  Conditions
     */
    const isNotGovCloud = new cdk.CfnCondition(this, "isNotGovCloud", {
        expression: cdk.Fn.conditionNot(cdk.Fn.conditionEquals(stack.partition, "aws-us-gov"))
    });
    {
        let childToMod = orchestratorLogGroupEncrypted.node.defaultChild as CfnLogGroup;
        childToMod.cfnOptions.condition = new cdk.CfnCondition(this, "Encrypted Log Group", {
            expression: cdk.Fn.conditionAnd(
                isNotGovCloud,
                cdk.Fn.conditionEquals(reuseOrchLogGroup.valueAsString, "no")
            )
        })
    }

    {
        let childToMod = orchestratorLogGroupNOTEncrypted.node.defaultChild as CfnLogGroup;
        childToMod.cfnOptions.condition = new cdk.CfnCondition(this, "Unencrypted Log Group", {
            expression: cdk.Fn.conditionAnd(
                cdk.Fn.conditionNot(isNotGovCloud),
                cdk.Fn.conditionEquals(reuseOrchLogGroup.valueAsString, "no")
            )
        })
        childToMod.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W84',
                    reason: 'KmsKeyId is not supported in GovCloud.'
                }]
            }
        }
    }
  }
} 