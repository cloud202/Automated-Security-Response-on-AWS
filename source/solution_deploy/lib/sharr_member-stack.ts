#!/usr/bin/env node
/*****************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/
import * as cdk_nag from 'cdk-nag';
import * as cdk from '@aws-cdk/core';
import * as fs from 'fs';
import { AdminAccountParm } from '../../lib/admin_account_parm-construct';
import { StringParameter } from '@aws-cdk/aws-ssm';
import * as ssm from '@aws-cdk/aws-ssm';
import * as s3 from '@aws-cdk/aws-s3';
import { Key } from '@aws-cdk/aws-kms';
import {
  PolicyStatement,
  Effect,
  PolicyDocument,
  ServicePrincipal
} from '@aws-cdk/aws-iam';
import { RunbookFactory } from './runbook_factory';
import * as lambda from '@aws-cdk/aws-lambda';
import { RemovalPolicy } from '@aws-cdk/core';

export interface SolutionProps {
  description: string;
  solutionId: string;
  solutionDistBucket: string;
  solutionTMN: string;
  solutionVersion: string;
  runtimePython: lambda.Runtime;
}

export class MemberStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props: SolutionProps) {
    super(scope, id, props);
    const stack = cdk.Stack.of(this);
    const RESOURCE_PREFIX = props.solutionId.replace(/^DEV-/, ''); // prefix on every resource name

    const adminAccount = new AdminAccountParm(this, 'AdminAccountParameter', {
      solutionId: props.solutionId
    });

    //Create a new parameter to track Redshift.4 S3 bucket
    const createS3BucketForRedshift4 = new cdk.CfnParameter(this, 'CreateS3BucketForRedshiftAuditLogging', {
      description: "Create S3 Bucket For Redshift Cluster Audit Logging.",
      type: "String",
      allowedValues:["yes", "no"],
      default: "no"
    });

    const enableS3BucketForRedShift4 = new cdk.CfnCondition(this,
      "EnableS3BucketForRedShift4",
      {
        expression: cdk.Fn.conditionEquals(createS3BucketForRedshift4.valueAsString, 'yes')
      })

    //Create the S3 Bucket for Redshift.4

    const s3BucketForAuditLogging = new s3.Bucket(this, "S3BucketForRedShiftAuditLogging", {
      encryption: s3.BucketEncryption.S3_MANAGED,
      publicReadAccess: false,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
    })

    cdk_nag.NagSuppressions.addResourceSuppressions(s3BucketForAuditLogging, [
      {id: 'AwsSolutions-S1', reason: 'This is a logging bucket.'}
    ]);

    const bucketPolicy = new s3.BucketPolicy(this, 'S3BucketForRedShiftAuditLoggingBucketPolicy', {
      bucket: s3BucketForAuditLogging,
      removalPolicy: RemovalPolicy.RETAIN
    })
    bucketPolicy.document.addStatements(
      new PolicyStatement({
        sid: 'Put bucket policy needed for audit logging',
        effect: Effect.ALLOW,
        actions: [
          "s3:GetBucketAcl",
          "s3:PutObject"
        ],
        principals: [new ServicePrincipal('redshift.amazonaws.com')],
        resources: [
          s3BucketForAuditLogging.bucketArn,
          cdk.Fn.sub("arn:${AWS::Partition}:s3:::${BucketName}/*", {
            BucketName: `${s3BucketForAuditLogging.bucketName}`
          })
        ]
      })
    );
    const bucketPolicy_cfn_ref = bucketPolicy.node.defaultChild as s3.CfnBucketPolicy
    bucketPolicy_cfn_ref.cfnOptions.condition = enableS3BucketForRedShift4

    const s3BucketForAuditLogging_cfn_ref = s3BucketForAuditLogging.node.defaultChild as s3.CfnBucket
    s3BucketForAuditLogging_cfn_ref.cfnOptions.metadata = {
      cfn_nag: {
        rules_to_suppress: [{
          id: 'W35',
          reason: 'Logs bucket does not require logging configuration'
        }]
      }
    };

    cdk_nag.NagSuppressions.addResourceSuppressions(s3BucketForAuditLogging, [
      {id: 'AwsSolutions-S1', reason: 'Logs bucket does not require logging configuration'},
      {id: 'AwsSolutions-S10', reason: 'Secure transport requirement is redundant for this use case'}
    ]);
    cdk_nag.NagSuppressions.addResourceSuppressions(bucketPolicy, [
      {id: 'AwsSolutions-S10', reason: 'Secure transport requirement is redundant for this use case'}
    ]);

    s3BucketForAuditLogging_cfn_ref.cfnOptions.condition = enableS3BucketForRedShift4;

    bucketPolicy_cfn_ref.addDependsOn(s3BucketForAuditLogging_cfn_ref)

    //--------------------------
    // KMS Customer Managed Key

    // Key Policy
    const kmsKeyPolicy:PolicyDocument = new PolicyDocument();
    const kmsPerms:PolicyStatement = new PolicyStatement();
    kmsPerms.addActions(
        'kms:GenerateDataKey',
        'kms:GenerateDataKeyPair',
        'kms:GenerateDataKeyPairWithoutPlaintext',
        'kms:GenerateDataKeyWithoutPlaintext',
        'kms:Decrypt',
        'kms:Encrypt',
        'kms:ReEncryptFrom',
        'kms:ReEncryptTo',
        'kms:DescribeKey',
        'kms:DescribeCustomKeyStores'
    );
    kmsPerms.effect = Effect.ALLOW;
    kmsPerms.addResources("*"); // Only the key the policydocument is attached to
    kmsPerms.addPrincipals(new ServicePrincipal('sns.amazonaws.com'));
    kmsPerms.addPrincipals(new ServicePrincipal('s3.amazonaws.com'));
    kmsPerms.addPrincipals(new ServicePrincipal(`logs.${stack.urlSuffix}`));
    kmsPerms.addPrincipals(new ServicePrincipal(`logs.${stack.region}.${stack.urlSuffix}`));
    kmsPerms.addPrincipals(new ServicePrincipal(`cloudtrail.${stack.urlSuffix}`));
    kmsPerms.addPrincipals(new ServicePrincipal('cloudwatch.amazonaws.com'));
    kmsKeyPolicy.addStatements(kmsPerms);

    const kmsKey:Key = new Key(this, 'SHARR Remediation Key', {
      enableKeyRotation: true,
      alias: `${RESOURCE_PREFIX}-SHARR-Remediation-Key`,
      trustAccountIdentities: true,
      policy: kmsKeyPolicy
    });

    new StringParameter(this, 'SHARR Key Alias', {
      description: 'KMS Customer Managed Key that will encrypt data for remediations',
      parameterName: `/Solutions/${RESOURCE_PREFIX}/CMK_REMEDIATION_ARN`,
      stringValue: kmsKey.keyArn
    });

    new StringParameter(this, 'SHARR Member Version', {
        description: 'Version of the AWS Security Hub Automated Response and Remediation solution',
        parameterName: `/Solutions/${RESOURCE_PREFIX}/member-version`,
        stringValue: props.solutionVersion
    });

    /********************
    ** Parameters
    ********************/

    const logGroupName = new cdk.CfnParameter(this, "LogGroupName", {
      type: "String",
      description: "Name of the log group to be used to create metric filters and cloudwatch alarms. You must use a Log Group that is the the logging destination of a multi-region CloudTrail"
    });

    /*********************************************
    ** Create SSM Parameter to store log group name
    *********************************************/
    new StringParameter(stack, 'SSMParameterLogGroupName', {
      description: 'Parameter to store log group name',
      parameterName: `/Solutions/${RESOURCE_PREFIX}/Metrics_LogGroupName`,
      stringValue: logGroupName.valueAsString
    });

    /*********************************************
    ** Create SSM Parameter to store encryption key alias for the PCI.S3.4/AFSBP.S3.4
    *********************************************/
    new StringParameter(stack, 'SSMParameterForS3.4EncryptionKeyAlias', {
      description: 'Parameter to store encryption key alias for the PCI.S3.4/AFSBP.S3.4, replace the default value with the KMS Key Alias, other wise the remediation will enable the default AES256 encryption for the bucket.',
      parameterName: `/Solutions/${RESOURCE_PREFIX}/afsbp/1.0.0/S3.4/KmsKeyAlias`,
      stringValue: 'default-s3-encryption'
    });

    /*********************************************
    ** Create SSM Parameter to store the S3 bucket name for AFSBP.REDSHIFT.4
    *********************************************/
    const ssmParameterForRedshift4BucketName = new StringParameter(stack, 'SSMParameterForS3BucketNameForREDSHIFT4', {
      description: 'Parameter to store the S3 bucket name for the remediation AFSBP.REDSHIFT.4, the default value is bucket-name which has to be updated by the user before using the remediation.',
      parameterName: `/Solutions/${props.solutionId}/afsbp/1.0.0/REDSHIFT.4/S3BucketNameForAuditLogging`,
      stringValue: s3BucketForAuditLogging.bucketName
    });

    const ssmParameterForRedshift4BucketName_cfn_ref = ssmParameterForRedshift4BucketName.node.defaultChild as ssm.CfnParameter
    ssmParameterForRedshift4BucketName_cfn_ref.cfnOptions.condition = enableS3BucketForRedShift4

    ssmParameterForRedshift4BucketName_cfn_ref.addDependsOn(s3BucketForAuditLogging_cfn_ref)


    new cdk.CfnMapping(this, 'SourceCode', {
      mapping: { "General": {
        "S3Bucket": props.solutionDistBucket,
        "KeyPrefix": props.solutionTMN + '/' + props.solutionVersion
      }}
    });

    const runbookFactory = new RunbookFactory(this, 'RunbookProvider', {
      solutionId: props.solutionId,
      runtimePython: props.runtimePython,
      solutionDistBucket: props.solutionDistBucket,
      solutionTMN: props.solutionTMN,
      solutionVersion: props.solutionVersion,
      region: this.region,
      partition: this.partition
    });

    //-------------------------------------------------------------------------
    // Runbooks - shared automations
    //
    const runbookStack = new cdk.CfnStack(this, `RunbookStackNoRoles`, {
      templateUrl: "https://" + cdk.Fn.findInMap("SourceCode", "General", "S3Bucket") +
      "-reference.s3.amazonaws.com/" + cdk.Fn.findInMap("SourceCode", "General", "KeyPrefix") +
      "/aws-sharr-remediations.template"
    });

    runbookStack.node.addDependency(runbookFactory);

    //-------------------------------------------------------------------------
    // Loop through all of the Playbooks to create reference
    //
    const PB_DIR = `${__dirname}/../../playbooks`;
    const ignore = ['.DS_Store', 'common', '.pytest_cache', 'NEWPLAYBOOK', '.coverage'];
    const illegalChars = /[\._]/g;
    const listOfPlaybooks: string[] = [];
    fs.readdir(PB_DIR, (err, items) => {
      items.forEach((file) => {
        if (!ignore.includes(file)) {
          const templateFile = `${file}MemberStack.template`;
          //---------------------------------------------------------------------
          // Playbook Member Template Nested Stack
          //
          const parmname = file.replace(illegalChars, '');
          const memberStackOption = new cdk.CfnParameter(this, `LoadMemberStack${parmname}`, {
            type: "String",
            description: `Load Playbook member stack for ${file}?`,
            default: "yes",
            allowedValues: ["yes", "no"]
          });
          memberStackOption.overrideLogicalId(`Load${parmname}MemberStack`);
          listOfPlaybooks.push(memberStackOption.logicalId);

          const memberStack = new cdk.CfnStack(this, `PlaybookMemberStack${file}`, {
            parameters: {
              'SecHubAdminAccount': adminAccount.adminAccountNumber.valueAsString
            },
            templateUrl: "https://" + cdk.Fn.findInMap("SourceCode", "General", "S3Bucket") +
            "-reference.s3.amazonaws.com/" + cdk.Fn.findInMap("SourceCode", "General", "KeyPrefix") +
            "/playbooks/" + templateFile
          })

          memberStack.node.addDependency(runbookFactory);

          memberStack.cfnOptions.condition = new cdk.CfnCondition(this, `load${file}Cond`, {
            expression: cdk.Fn.conditionEquals(memberStackOption, "yes")
          });
        }
      });
    });
    /********************
    ** Metadata
    ********************/
    stack.templateOptions.metadata = {
      "AWS::CloudFormation::Interface": {
        ParameterGroups: [
          {
            Label: {default: "LogGroup Configuration"},
            Parameters: [logGroupName.logicalId]
          },
          {
            Label: {default: "Playbooks"},
            Parameters: listOfPlaybooks
          }
        ],
        ParameterLabels: {
          [logGroupName.logicalId]: {
            default: "Provide the name of the LogGroup to be used to create Metric Filters and Alarms",
          }
        }
      }
    };
  }
}
