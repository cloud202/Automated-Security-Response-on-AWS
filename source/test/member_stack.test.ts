/*****************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import { SynthUtils } from '@aws-cdk/assert';
import * as lambda from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';
import { MemberStack } from '../solution_deploy/lib/sharr_member-stack';
import { AwsSolutionsChecks } from 'cdk-nag'
import { Aspects } from '@aws-cdk/core'

function getCatStack(): cdk.Stack {
  const app = new cdk.App();
  const stack = new MemberStack(app, 'CatalogStack', {
    description: 'test;',
    solutionId: 'SO0111',
    solutionVersion: 'v1.1.1',
    solutionDistBucket: 'sharrbukkit',
    solutionTMN: 'aws-security-hub-automated-response-and-remediation',
    runtimePython: lambda.Runtime.PYTHON_3_8
  });
  Aspects.of(app).add(new AwsSolutionsChecks({verbose: true}))
  return stack;
}

test('default stack', () => {
  expect(SynthUtils.toCloudFormation(getCatStack())).toMatchSnapshot();
});


