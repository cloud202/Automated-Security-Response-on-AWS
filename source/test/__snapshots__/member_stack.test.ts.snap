// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default stack 1`] = `
Object {
  "Conditions": Object {
    "EnableS3BucketForRedShift4": Object {
      "Fn::Equals": Array [
        Object {
          "Ref": "CreateS3BucketForRedshiftAuditLogging",
        },
        "yes",
      ],
    },
  },
  "Description": "test;",
  "Mappings": Object {
    "SourceCode": Object {
      "General": Object {
        "KeyPrefix": "aws-security-hub-automated-response-and-remediation/v1.1.1",
        "S3Bucket": "sharrbukkit",
      },
    },
  },
  "Metadata": Object {
    "AWS::CloudFormation::Interface": Object {
      "ParameterGroups": Array [
        Object {
          "Label": Object {
            "default": "LogGroup Configuration",
          },
          "Parameters": Array [
            "LogGroupName",
          ],
        },
        Object {
          "Label": Object {
            "default": "Playbooks",
          },
          "Parameters": Array [],
        },
      ],
      "ParameterLabels": Object {
        "LogGroupName": Object {
          "default": "Provide the name of the LogGroup to be used to create Metric Filters and Alarms",
        },
      },
    },
  },
  "Parameters": Object {
    "CreateS3BucketForRedshiftAuditLogging": Object {
      "AllowedValues": Array [
        "yes",
        "no",
      ],
      "Default": "no",
      "Description": "Create S3 Bucket For Redshift Cluster Audit Logging.",
      "Type": "String",
    },
    "LogGroupName": Object {
      "Description": "Name of the log group to be used to create metric filters and cloudwatch alarms. You must use a Log Group that is the the logging destination of a multi-region CloudTrail",
      "Type": "String",
    },
    "SecHubAdminAccount": Object {
      "AllowedPattern": "\\\\d{12}",
      "Description": "Admin account number",
      "Type": "String",
    },
  },
  "Resources": Object {
    "RunbookProviderFunction82CD9D9B": Object {
      "DependsOn": Array [
        "RunbookProviderRoleBC4E91CA",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "False positive. Access is provided via a policy",
            },
            Object {
              "id": "W89",
              "reason": "There is no need to run this lambda in a VPC",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Join": Array [
              "",
              Array [
                "sharrbukkit-",
                Object {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          "S3Key": "aws-security-hub-automated-response-and-remediation/v1.1.1/lambda/updatableRunbookProvider.py.zip",
        },
        "Description": "Custom resource to manage versioned SSM documents",
        "Environment": Object {
          "Variables": Object {
            "LOG_LEVEL": "info",
            "SOLUTION_ID": "AwsSolution/SO0111/v1.1.1",
          },
        },
        "FunctionName": "SO0111-SHARR-updatableRunbookProvider",
        "Handler": "updatableRunbookProvider.lambda_handler",
        "Layers": Array [
          Object {
            "Ref": "RunbookProviderMemberLambdaLayerF3BD824A",
          },
        ],
        "MemorySize": 256,
        "ReservedConcurrentExecutions": 1,
        "Role": Object {
          "Fn::GetAtt": Array [
            "RunbookProviderRoleBC4E91CA",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 600,
      },
      "Type": "AWS::Lambda::Function",
    },
    "RunbookProviderMemberLambdaLayerF3BD824A": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "python3.8",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Fn::Join": Array [
              "",
              Array [
                "sharrbukkit-",
                Object {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          "S3Key": "aws-security-hub-automated-response-and-remediation/v1.1.1/lambda/memberLayer.zip",
        },
        "Description": "SO0111 SHARR Common functions used by the solution member stack",
        "LicenseInfo": "https://www.apache.org/licenses/LICENSE-2.0",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "RunbookProviderPolicy12B46DCD": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "Resource * is required in order to manage arbitrary SSM documents",
            },
          ],
        },
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Resource * is required in order to manage arbitrary SSM documents",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "ssm:CreateDocument",
                "ssm:UpdateDocument",
                "ssm:UpdateDocumentDefaultVersion",
                "ssm:ListDocumentVersions",
                "ssm:DeleteDocument",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SO0111-SHARR_Runbook_Provider_Policy",
        "Roles": Array [
          Object {
            "Ref": "RunbookProviderRoleBC4E91CA",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RunbookProviderRoleBC4E91CA": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Lambda role to allow creation of updatable SSM documents",
      },
      "Type": "AWS::IAM::Role",
    },
    "RunbookStackNoRoles": Object {
      "DependsOn": Array [
        "RunbookProviderFunction82CD9D9B",
        "RunbookProviderMemberLambdaLayerF3BD824A",
        "RunbookProviderPolicy12B46DCD",
        "RunbookProviderRoleBC4E91CA",
      ],
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://",
              Object {
                "Fn::FindInMap": Array [
                  "SourceCode",
                  "General",
                  "S3Bucket",
                ],
              },
              "-reference.s3.amazonaws.com/",
              Object {
                "Fn::FindInMap": Array [
                  "SourceCode",
                  "General",
                  "KeyPrefix",
                ],
              },
              "/aws-sharr-remediations.template",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
    },
    "S3BucketForRedShiftAuditLogging652E7355": Object {
      "Condition": "EnableS3BucketForRedShift4",
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-S1",
              "reason": "Logs bucket does not require logging configuration",
            },
            Object {
              "id": "AwsSolutions-S10",
              "reason": "Secure transport requirement is redundant for this use case",
            },
          ],
        },
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W35",
              "reason": "Logs bucket does not require logging configuration",
            },
          ],
        },
      },
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "S3BucketForRedShiftAuditLoggingBucketPolicyAB8BAA40": Object {
      "Condition": "EnableS3BucketForRedShift4",
      "DeletionPolicy": "Retain",
      "DependsOn": Array [
        "S3BucketForRedShiftAuditLogging652E7355",
      ],
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-S10",
              "reason": "Secure transport requirement is redundant for this use case",
            },
          ],
        },
      },
      "Properties": Object {
        "Bucket": Object {
          "Ref": "S3BucketForRedShiftAuditLogging652E7355",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetBucketAcl",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "Service": "redshift.amazonaws.com",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "S3BucketForRedShiftAuditLogging652E7355",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Sub": Array [
                    "arn:\${AWS::Partition}:s3:::\${BucketName}/*",
                    Object {
                      "BucketName": Object {
                        "Ref": "S3BucketForRedShiftAuditLogging652E7355",
                      },
                    },
                  ],
                },
              ],
              "Sid": "Put bucket policy needed for audit logging",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
      "UpdateReplacePolicy": "Retain",
    },
    "SHARRKeyAliasEBF509D8": Object {
      "Properties": Object {
        "Description": "KMS Customer Managed Key that will encrypt data for remediations",
        "Name": "/Solutions/SO0111/CMK_REMEDIATION_ARN",
        "Type": "String",
        "Value": Object {
          "Fn::GetAtt": Array [
            "SHARRRemediationKeyE744743D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SHARRMemberVersionEDAB5C42": Object {
      "Properties": Object {
        "Description": "Version of the AWS Security Hub Automated Response and Remediation solution",
        "Name": "/Solutions/SO0111/member-version",
        "Type": "String",
        "Value": "v1.1.1",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SHARRRemediationKeyAlias5531874D": Object {
      "Properties": Object {
        "AliasName": "alias/SO0111-SHARR-Remediation-Key",
        "TargetKeyId": Object {
          "Fn::GetAtt": Array [
            "SHARRRemediationKeyE744743D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
    "SHARRRemediationKeyE744743D": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "EnableKeyRotation": true,
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kms:GenerateDataKey",
                "kms:GenerateDataKeyPair",
                "kms:GenerateDataKeyPairWithoutPlaintext",
                "kms:GenerateDataKeyWithoutPlaintext",
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncryptFrom",
                "kms:ReEncryptTo",
                "kms:DescribeKey",
                "kms:DescribeCustomKeyStores",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "sns.amazonaws.com",
                  "s3.amazonaws.com",
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "logs.",
                        Object {
                          "Ref": "AWS::URLSuffix",
                        },
                      ],
                    ],
                  },
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "logs.",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ".",
                        Object {
                          "Ref": "AWS::URLSuffix",
                        },
                      ],
                    ],
                  },
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "cloudtrail.",
                        Object {
                          "Ref": "AWS::URLSuffix",
                        },
                      ],
                    ],
                  },
                  "cloudwatch.amazonaws.com",
                ],
              },
              "Resource": "*",
            },
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "SSMParameterForS34EncryptionKeyAlias73DD8A98": Object {
      "Properties": Object {
        "Description": "Parameter to store encryption key alias for the PCI.S3.4/AFSBP.S3.4, replace the default value with the KMS Key Alias, other wise the remediation will enable the default AES256 encryption for the bucket.",
        "Name": "/Solutions/SO0111/afsbp/1.0.0/S3.4/KmsKeyAlias",
        "Type": "String",
        "Value": "default-s3-encryption",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SSMParameterForS3BucketNameForREDSHIFT441DD36B1": Object {
      "Condition": "EnableS3BucketForRedShift4",
      "DependsOn": Array [
        "S3BucketForRedShiftAuditLogging652E7355",
      ],
      "Properties": Object {
        "Description": "Parameter to store the S3 bucket name for the remediation AFSBP.REDSHIFT.4, the default value is bucket-name which has to be updated by the user before using the remediation.",
        "Name": "/Solutions/SO0111/afsbp/1.0.0/REDSHIFT.4/S3BucketNameForAuditLogging",
        "Type": "String",
        "Value": Object {
          "Ref": "S3BucketForRedShiftAuditLogging652E7355",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SSMParameterLogGroupName47918519": Object {
      "Properties": Object {
        "Description": "Parameter to store log group name",
        "Name": "/Solutions/SO0111/Metrics_LogGroupName",
        "Type": "String",
        "Value": Object {
          "Ref": "LogGroupName",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;
